Extend the fusion/e2e tests to f16 and bf16.

pcf000 opened the pull request:
Modify most of the tests in fusion/e2e and fusion/e2e/mixr-gemm to make them into templates, into which the generate-fusion-tests.py script will substitute the element type and a couple of other fields.

In the general case, only the type is varied, plus the clone-verification run is disabled if the type isn't f32, because its underlying mcpuVerify() doesn't support f16 and bf16 yet.

Several tosa-to-rock tests are built from one template, because the test differs only in a single op;  in those cases, type, op, and a couple supporting fields are varied.

The script is run by cmake, and there's a lot of boilerplate in there to connect it to the check-rocmlir target and to make sure that its environment is correct.

pcf000 committed:
Extend the fusion/e2e tests to f16 and bf16.
Modify most of the tests in fusion/e2e and fusion/e2e/mixr-gemm to make
them into templates, into which the generate-fusion-tests.py script will
substitute the element type and a couple of other fields.

In the general case, only the type is varied, plus the
clone-verification run is disabled if the type isn't f32, because its
underlying mcpuVerify() doesn't support f16 and bf16 yet.

Several tosa-to-rock tests are built from one template, because the test
differs only in a single op;  in those cases, type, op, and a couple
supporting fields are varied.

The script is run by cmake, and there's a lot of boilerplate in there to
connect it to the check-rocmlir target and to make sure that its
environment is correct.

Move fusion/e2e add-subdirectory under ROCK_E2E_TEST_ENABLED to avoid
dependence error when -DBUILD_FAT_LIBROCKCOMPILER=ON.

Copy the non-templated *.mlir files into the build directory, too,
because cmake isn't.

New and improved:  template files use f-string syntax, tests specified
in .toml, cleaner test generation.

xmodel changed to mhal.

Change condition in lit.local.cfg.

Copy lit.local.cfg while generating tests.

pcf000 commented on mlir/test/CMakeLists.txt:
```diff
@@ -36,6 +37,7 @@ message(STATUS "MLIR_CONV_VALIDATION_WRAPPER_LIBRARY_DIR: ${MLIR_CONV_VALIDATION
 # Need to explicitly give paths to llvm tools
 set(LLVM_LIT_TOOLS_DIR ${LLVM_EXTERNAL_BIN_DIR})
 set(LLVM_EXTERNAL_LIT "${LLVM_LIT_TOOLS_DIR}/llvm-lit")
+set(LLVM_DEFAULT_EXTERNAL_LIT "${LLVM_LIT_TOOLS_DIR}/llvm-lit")
```
I need this variable to ensure it looks for llvm-lit in the external bin directory and not build/bin.

krzysz00 replied:
Do you still need this?

krzysz00 reviewed (commented):
Comments:
1. Why didn't the `e2e` infrastructure need the LLVM_EXTERNAL_LIT variable? Maybe there's something in that site.cfg that's relevant.
2. Can we use the template functionality more / could the Python script do more work? That is, for instance, why not have the `// RUN` lines autogenerated? Or at least mostly filled in, so we don't have a million copies of the shared-libs invocation to update if we ever change that?
3. Do the generated files live over in `build/` just like the autogenerated e2e tests?

zhanglx13 commented on mlir/test/CMakeLists.txt:
```diff
@@ -142,3 +145,11 @@ configure_lit_site_cfg(
   ${CMAKE_CURRENT_SOURCE_DIR}/e2e/lit.cfg.py
 )
+
+# Update the configs for E2E tests
```
Can you add fusion in the comment so that we can distinguish fusion e2e test from the rock e2e tests?

pcf000 replied:
Done.

zhanglx13 commented on mlir/test/fusion/e2e/CMakeLists.txt:
```diff
@@ -0,0 +1,37 @@
+  mlir-cpu-runner
+  mlir_rocm_runtime
+  conv-validation-wrappers
+  e2e_tests
```
I think we shouldn't include e2e_tests here.

pcf000 replied:
Needed to be fusion-e2e-tests instead.

zhanglx13 commented on mlir/test/fusion/e2e/CMakeLists.txt:
```diff
@@ -0,0 +1,37 @@
+  )
+
+# Add a new target to run the E2E test only
+add_lit_testsuite(check-fusion-e2e "Running fusion e2e tests"
```
So I tried `ninja check-fusion-e2e` and I got the error:
/bin/sh: 1: /home/zhanglx/rocMLIR/build/bin/llvm-lit: not found
Can you check if this is the case on your end?
This the cmake command I used before running ninja
```
cmake -G Ninja -D CMAKE_BUILD_TYPE=RelWithDebInfo -DROCMLIR_DRIVER_ENABLED=1 -DROCMLIR_DRIVER_PR_E2E_TEST_ENABLED=1 -DROCMLIR_DRIVER_E2E_TEST_ENABLED=0 -DROCK_E2E_TEST_ENABLED=0 -DROCMLIR_DRIVER_TEST_GPU_VALIDATION=1 "-DLLVM_LIT_ARGS=-v --time-tests" -DCMAKE_EXPORT_COMPILE_COMMANDS=1 ..
```

pcf000 replied:
Hmm.  That's the whole reason I fiddled with LLVM_DEFAULT_EXTERNAL_LIT and LLVM_EXTERNAL_LIT.  I will see if I can reproduce the problem on a clean environment.

pcf000 replied:
Fixed this error.  Ultimately, it was because "add_subdirectory(fusion)" came before the setting of LLVM_LIT_TOOLS_DIR et al in mlir/test/CMakeLists.txt.

zhanglx13 commented on mlir/test/fusion/e2e/CMakeLists.txt:
```diff
@@ -0,0 +1,37 @@
+set(LOCAL_CONFIG_FILE_BIN ${CMAKE_CURRENT_BINARY_DIR}/lit.local.cfg)
+
+# Add a command to copy lit.local.cfg
+add_custom_command(OUTPUT ${LOCAL_CONFIG_FILE_BIN}
```
It seems that the fusion e2e tests are not guarded by any macros. So you do not need a `lit.local.cfg` to control when to run the tests. 
And this `add_custon_command()` does not have any effect because `${LOCAL_CONFIG_FILE_BIN}` is not a target.

pcf000 replied:
I copied a lot of the cmake files from other test sets.  I think I need to reexamine them.

pcf000 replied:
Ultimately I did need a lit.local.cfg, and this custom-command still didn't do anything, and now I copy lit.local.cfg in generate-fusion-tests.py.

krzysz00 replied:
That shouldn't be it.

You might need to depend on the copy, or on the output, or otherwise get the command into CMake's dependency chain.

krzysz00 commented on mlir/test/CMakeLists.txt:
```diff
@@ -109,6 +110,7 @@ if(ROCMLIR_DRIVER_ENABLED)
     mlir-translate
     opt
     llc
+    fusion-e2e-tests
```
Wrong list of dependencies. You want this handled the same as the `e2e` test directory.

pcf000 replied:
Well, it's unconditionally add-subdirectory'd further down.  Would you prefer that it be APPENDed down there, or that it be underneath the ROCK_E2E_TEST_ENABLED conditional like the e2e directory?

krzysz00 replied:
I'd move it into the conditional. Heck, I'd move the e2e fusion tests under `e2e/`, thus needing fewer lit configs.

pcf000 replied:
Turns out that making it conditional is also required to make -DBUILD_FAT_LIBROCKCOMPILER=ON work.

krzysz00 reviewed (commented):
Minor note requesting guarding these tests via a similar method to  the `e2e` directory. Heck, maybe move these to `e2e/fusion`

pcf000 commented:
Moved the "add_subdirectory(fusion)" under the conditional with the "add_subdirectory(e2e)".  Not only did Krzysztof prefer it, but it also was needed to avoid problems with -DBUILD_FAT_LIBROCKCOMPILER=ON.

Added some logic to generate-fusion-tests.py to copy over the non-templated *.mlir files (tests for which multiple data types were infeasible or irrelevant), because that wasn't happening any more.  Those tests required a couple more dependences.

krzysz00 commented on mlir/test/CMakeLists.txt:
```diff
@@ -114,10 +115,12 @@ endif()
-  list(APPEND ROCMLIR_TEST_DEPENDS e2e_tests)
+  list(APPEND ROCMLIR_TEST_DEPENDS e2e_tests fusion-e2e-tests)
   add_subdirectory(e2e)
+  add_subdirectory(fusion)
```
I'm not liking this. There are non e2e fusion tests.

Could you instead refactor the tests from `fusion/e2e` to `e2e/fusion`

pcf000 replied:
There are also non-fusion e2e tests.  I was leaving the original arrangement, but I'll consider moving them.

krzysz00 commented on mlir/test/fusion/e2e/generate-fusion-tests.py:
```diff
@@ -0,0 +1,91 @@
+parser.add_argument('outdir', default=os.getcwd())
+args = parser.parse_args()
+
+def run():
```
This feels ... really fragile and hard to discover.

Would you be open to using a configuration file like we use in the main e2e tests. TOML exists, and we could use it to express what test suites we have, what directories they're in, and how we want them parameterized.

pcf000 replied:
Not sure what you have in mind.  I think you're asking that I recast the .template files as files containing f-strings (which is probably clearer anyway), and also change some of the assignments and parameters in the generate.py file into a .toml file.

The e2e tests are simpler in that they have the entire text of the tests within the generate.py, since they're only about three lines long and easily filled in.  The fusion/e2e tests are varied .mlir files;  including the .template text in generate.py sounds like going too far.

pcf000 replied:
Redid the test generation to use a .toml file and f-string-like template files.  Makes some things simpler, but requires a workaround for the operand to tosa.clamp.

krzysz00 reviewed (commented):
I've got some comments about the overall approach here.

krzysz00 commented on mlir/test/fusion/e2e/generate-fusion-tests.py:
```diff
@@ -0,0 +1,100 @@
+    for suite in toml['suite']:
+        combinations = generate_option_list(suite, 'axis', 'values')
+        for test in combinations:
+            if suite['name'] == "tosa-to-rock tests":
```
Instead of keying off the name field, would you be open to adding a `type` field to `suite`

pcf000 replied:
Okay, but I called it "kind" because "type" has so many other uses.

krzysz00 reviewed (commented):
Minor note about the template generation here

pcf000 committed:
Make a 'kind' field in the .toml file to distinguish the two kinds of test generation.

zhanglx13 commented on mlir/test/fusion/e2e/lit.site.cfg.py.in:
```diff
@@ -0,0 +1,75 @@
+        agents = set(x.decode("utf-8") for x in p.stdout.split())
+        if not agents:
+            # TODO: Remove this workaround for a bug in rocm_agent_enumerator -name
+            # Once https://github.com/RadeonOpenCompute/rocminfo/pull/59 lands
```
someone said it's merged in the internal repo. Does it mean we need to remove this workaround?

zhanglx13 reviewed (approved):
LGTM.
Just one comment about the workaround for rocminfo. @krzysz00

manupak commented:
@jerryyin @pcf000 this PR woudn;t have passed CI if navi3x was not disabled && if we use gfx1101 node in the CI.

pcf000 commented:
@manupak , could you explain?

manupak commented:
http://rocmhead:8080/blue/organizations/jenkins/MLIR%2Fmlir/detail/PR-1248/4/pipeline/172

Here is where I found about this.

Then manually verified on ctr-smc-s39-007 (gfx1101) node using the current develop. 
What amused me is it is passing on gfx1100
